cpp/src/arrow/util/heap.h

```diff
@@ -0,0 +1,76 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <algorithm>
+#include <functional>
+#include <vector>
+#include "arrow/util/macros.h"
+
+namespace arrow {
+namespace internal {
+
+// A Heap class, is a simple wrapper to make heap operation simpler.
+// This class is immutable by design
+template <typename T, typename Compare = std::less<T>>
+class ARROW_EXPORT Heap {
+ public:
+  explicit Heap() : values_(), comp_() {}
+  explicit Heap(const Compare& compare) : values_(), comp_(compare) {}
+
+  Heap(Heap&&) = default;
+  Heap& operator=(Heap&&) = default;
+
+  T* Data() { return values_.data(); }
+
+  // const T& Top() const { return values_.front(); }
+
+  T Top() const { return values_.front(); }
+
+  bool Empty() const { return values_.empty(); }
+
+  size_t Size() const { return values_.size(); }
```

Comment:
nit: for const getter methods like this, they can be named lowercase still. size() is often left as size() and not Size() in our codebase, and top()/empty()/data() could also follow that.